serialization: manage uncompressed point serialization for trusted sources

jsign opened the pull request:
This PR is an experimental change to avoid persisting compressed EC points in the resolver. 
This will allow parsing nodes to use less CPU since no `Y` coordinate should be calculated from `X`.

While doing this change, I switched multiple places with "magic numbers" with more explicit calculations and constants.

This PR is using a [custom go-ipa version](https://github.com/crate-crypto/go-ipa/pull/33).

jsign commented on encoding.go:
```diff
@@ -44,21 +44,21 @@ func bit(bitlist []byte, nr int) bool {
 
 func ParseNode(serialized []byte, depth byte, comm []byte) (VerkleNode, error) {
-	if len(serialized) < 64 {
+	if len(serialized) < 1+StemSize+SerializedAffinePointSize {
```
Transforming magic numbers into more obvious constant calculations. This wasn't trivial now that the size of the affine point changed it's size.

jsign commented on encoding.go:
```diff
@@ -44,21 +44,21 @@ func bit(bitlist []byte, nr int) bool {
 		return parseLeafNode(serialized, depth, comm)
 	case internalRLPType:
-		return CreateInternalNode(serialized[1:33], serialized[33:], depth, comm)
+		return CreateInternalNode(serialized[1:1+NodeWidth/8], serialized[1+NodeWidth/8:], depth, comm)
```
`NodeWith/8` is a clearer calculation to signal the bitfield size.

jsign commented on encoding.go:
```diff
@@ -86,12 +86,12 @@ func parseLeafNode(serialized []byte, depth byte, comm []byte) (VerkleNode, erro
 	if NodeWidth != len(values) {
```
All these are doing two things at once:
- Avoiding more magic numbers
- Indirectly the `SerializedAffinePointSize` has been switched from 32 to 64, since now both points are part of the serialization.

jsign commented on encoding_test.go:
```diff
@@ -18,7 +18,8 @@ func TestLeafStemLength(t *testing.T) {
 		t.Fatal(err)
 	}
-	if len(ser) != 128 {
+	// RLP_type + Stem + bitset + 2-commitments
+	if len(ser) != 1+StemSize+NodeWidth/8+2*SerializedAffinePointSize {
```
Went a bit further with a comment to explain this that wasn't trivial.

jsign commented on go.mod:
```diff
@@ -5,3 +5,5 @@ go 1.18
 
 require golang.org/x/sys v0.0.0-20220919091848-fb04ddd9f9c8 // indirect
+
+replace github.com/crate-crypto/go-ipa => github.com/jsign/go-ipa v0.0.0-20230131214742-21080f1193c2
```
@gballet, I had to do this since the `go-ipa-` branch in https://github.com/crate-crypto/go-ipa/pull/33 isn't part of the `go-ipa` repo.

If you're going to test this in the benchmark, you'll have to add this `replace` line in the `go-ethereum` `go.mod` file since `replaces` aren't "inherited" in dependencies.

jsign commented on ipa.go:
```diff
@@ -31,10 +31,13 @@ import (
-	Point = banderwagon.Element
+	Fr                    = fr.Element
+	Point                 = banderwagon.Element
+	SerializedAffinePoint = banderwagon.SerializedAffinePoint
 )
 
+const SerializedAffinePointSize = banderwagon.SerializedAffinePointSize
```
New type and constant from go-ipa that are pulled here to be more explicit and type safe.

jsign commented on proof_ipa.go:
```diff
@@ -138,7 +138,7 @@ func SerializeProof(proof *Proof) ([]byte, []KeyValuePair, error) {
 
```
Remember that for proofs, we still use compressed serialization.

gballet replied:
proofs and commitments computation alike. With this approach, I get a different root commitment, so something must impact commitment calculation, when it shouldn't.

jsign replied:
All tests, proof gen, and verification still pass.
One of the tests is checking compatibility with Rust and is still passing.

This sounds more like a `go-ethereum` call to the changed semantics `Bytes()` as we chatted; we can look at it.

jsign commented:
@gballet, I'm closing this PR, and I'll open a separate one with these changes plus other optimizations in a separate one.
I'll be addressing all your comments in this PR in the new version, so comments/feedback aren't lost.
